compiler/rustc_const_eval/src/interpret/projection.rs

```diff
@@ -37,31 +38,32 @@ pub trait Projectable<'tcx, Prov: Provenance>: Sized + std::fmt::Debug {
     }
 
     /// Offset the value by the given amount, replacing the layout and metadata.
-    fn offset_with_meta(
+    fn offset_with_meta<'mir, M: Machine<'mir, 'tcx, Provenance = Prov>>(
         &self,
         offset: Size,
         meta: MemPlaceMeta<Prov>,
         layout: TyAndLayout<'tcx>,
-        cx: &impl HasDataLayout,
+        ecx: &InterpCx<'mir, 'tcx, M>,
```

Comment:
My plan is to add a `OffsetMode` enum parameter controlling whether we use inbounds or nowrap (or maybe wrapping) offset.